### PR TITLE
Increase test coverage in backend

### PR DESCRIPTION
**Pull request by holgerkoser**
**What this PR does / why we need it**:
With this PullRequest the test coverage is significantly improved and many parts of the code that where untested before are now tested.

Screen Shot 2020-01-29 at 16 28 23

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```improvement developer
Increase test coverage in backend significantly
```

### CODE REVIEW

File: backend/server.js [hunk 1]
```diff
@@ -16,54 +16,9 @@
...
-  logger.debug('Cleanup has been finished. Server is shutting down')
-}
-
-function toMilliseconds (seconds) {
-  seconds = parseInt(seconds)
-  if (isNaN(seconds)) {
-    seconds = 10
-  }
-  return seconds * 1000 + 200
+if (require.main === module) {
```
Comment (petersutter) [inline backend/server.js]: comment or rm

File: backend/lib/server.js [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    }
+  })
+
+  server.startListeningForConnections = function startListeningForConnections () {
+    const startServer = async () => {
+      await new Promise(resolve => server.listen(port, resolve))
+      logger.info(`Server listening on port ${port}`)
+    }
+    const seconds = 30
+    return pTimeout(startServer(), seconds * 1000, `Starting server timed out after ${seconds} seconds`)
```
Comment (petersutter) [inline backend/lib/server.js]: is wrapping server.listen really needed?

File: backend/lib/watches/journals.js [hunk 3]
```diff
@@ -45,26 +45,26 @@ module.exports = io => {
...
-    loadOpenIssues().then(() => cb(), err => cb(err))
-  }
-
-  const call = backoff.call(loadAllOpenIssues, err => {
-    if (err) {
-      logger.error('failed to fetch journals', err)
-    } else {
+  async function loadAllOpenIssues () {
+    const options = {
+      retries: 0,
```
Comment (petersutter) [inline backend/lib/watches/journals.js]: why retries: 0 and forever: true?

File: backend/test/support/nocks/k8s.js [hunk 4]
```diff
@@ -811,27 +817,84 @@ const stub = {
...
+    if (target === 'cp') {
+      scope
+        .get(`/apis/core.gardener.cloud/v1alpha1/namespaces/garden/shoots/${seedName}`)
+        .reply(200, {
+          metadata: {
+            name: seedName,
+            namespace: 'garden'
+          },
+          status: {
+            seed: 'soil-infra1'
```
Comment (petersutter) [inline backend/test/support/nocks/k8s.js]: seedName in spec

File: backend/test/services.terminals.spec.js [hunk 5]
```diff
@@ -0,0 +1,577 @@
...
+    describe('terminals', function () {
+      describe('#getGardenTerminalHostClusterSecretRef', function () {
+        it('should return the secret reference by secretRef', async function () {
+          const gardenTerminalHost = {
+            secretRef: {
+              namespace: 'secretNamespace'
+            }
+          }
+          createConfigStub({ gardenTerminalHost })
+          const secretRef = await getGardenTerminalHostClusterSecretRef(client)
```
Comment (petersutter) [inline backend/test/services.terminals.spec.js]: verify labelselector

File: backend/test/services.terminals.spec.js [hunk 6]
```diff
@@ -0,0 +1,577 @@
...
+          },
+          subsets
+        })
+      })
+    })
+
+    describe('bootstrap', function () {
+      const hostClient = {
+        cluster: {
+          server: new URL('https://apiServerHostname:6443')
```
Comment (petersutter) [inline backend/test/services.terminals.spec.js]: different server urls

File: backend/test/services.terminals.spec.js [hunk 7]
```diff
@@ -0,0 +1,577 @@
...
+          }
+          if (name === `seedsecret-${seedName}`) {
+            return seedClient
+          }
+        })
+      })
+
+      it('should not bootstrap anything', async function () {
+        const bootstrap = {
+          disabled: true
```
Comment (petersutter) [inline backend/test/services.terminals.spec.js]: in addition, bootstrapResource seed and shoot -> should not be added to queue

File: backend/test/utils.spec.js [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+    it('should return the seed name for a shoot resource', function () {
+      expect(() => getSeedNameFromShoot({})).to.throw(AssertionError)
+      const shoot = {
+        spec: {
+          seedName: 'foo'
+        },
+        status: {}
+      }
+      expect(getSeedNameFromShoot(shoot)).to.equal('foo')
+      shoot.status.seed = 'bar'
```
Comment (petersutter) [inline backend/test/utils.spec.js]: rm

File: backend/test/watches.spec.js [hunk 9]
```diff
@@ -0,0 +1,527 @@
...
+      watches.cloudprofiles(io)
+      expect(watchStub).to.be.calledOnce
+      expect(cacheStub).to.be.calledOnce
+      items.push(foo)
+      emitter.emit('connect')
+      expect(items).to.be.empty
+      emitter.emit('event', { type: 'DELETED', object: foo })
+      emitter.emit('event', { type: 'ADDED', object: foo })
+      emitter.emit('event', { type: 'ADDED', object: bar })
+      emitter.emit('event', { type: 'MODIFIED', object: { kind, ...bar } })
```
Comment (petersutter) [inline backend/test/watches.spec.js]: expect 2 items

File: backend/test/watches.spec.js [hunk 10]
```diff
@@ -0,0 +1,527 @@
...
+        return nsp
+      }
+    }
+    const { projects } = dashboardClient['core.gardener.cloud']
+
+    it('should watch projects', async function () {
+      const watchStub = sandbox.stub(projects, 'watchList').returns(emitter)
+      watches.namespaces(io)
+      expect(watchStub).to.be.calledOnce
+      emitter.emit('event', { type: 'ADDED', object: foo })
```
Comment (petersutter) [inline backend/test/watches.spec.js]: is something missing here?

File: backend/test/watches.spec.js [hunk 11]
```diff
@@ -0,0 +1,527 @@
...
+
+    const io = {
+      of (namespace) {
+        expect(namespace).to.equal('/shoots')
+        return nsp
+      }
+    }
+    const { projects } = dashboardClient['core.gardener.cloud']
+
+    it('should watch projects', async function () {
```
Comment (petersutter) [inline backend/test/watches.spec.js]: delete project watch coding

File: backend/test/watches.spec.js [hunk 12]
```diff
@@ -0,0 +1,527 @@
...
+      cacheStub.returns(journalCache)
+      loadOpenIssuesStub.onCall(0).throws(serviceUnavailable)
+      watches.journals(io, { minTimeout: 1 })
+      expect(cacheStub).to.be.calledOnce
+      await delay(5)
+      expect(loadOpenIssuesStub).to.be.calledTwice
+      expect(infoSpy).to.be.calledTwice
+    })
+
+    it('should failed to fetch journals', async function () {
```
Comment (petersutter) [inline backend/test/watches.spec.js]:
```suggestion
    it('should fail to fetch journals', async function () {
```

File: backend/test/watches.spec.js [hunk 13]
```diff
@@ -0,0 +1,527 @@
...
+      expect(warnSpy).to.be.calledOnce
+    })
+
+    it('should watch journals', async function () {
+      gitHubConfigStub.get(() => true)
+      cacheStub.returns(journalCache)
+      loadOpenIssuesStub.onCall(0).throws(serviceUnavailable)
+      watches.journals(io, { minTimeout: 1 })
+      expect(cacheStub).to.be.calledOnce
+      await delay(5)
```
Comment (petersutter) [inline backend/test/watches.spec.js]: use explicit finish event instead of fixed delay

File: backend/test/watches.spec.js [hunk 14]
```diff
@@ -0,0 +1,527 @@
...
+
+    it('should failed to fetch journals', async function () {
+      gitHubConfigStub.get(() => true)
+      cacheStub.returns(journalCache)
+      loadOpenIssuesStub.throws(new Error('Unexpected'))
+
+      watches.journals(io)
+      expect(cacheStub).to.be.calledOnce
+      expect(loadOpenIssuesStub).to.be.calledOnce
+      await nextTick()
```
Comment (petersutter) [inline backend/test/watches.spec.js]: use explicit finish event

File: backend/test/acceptance/io.spec.js [hunk 15]
```diff
@@ -0,0 +1,282 @@
...
+      { id: 2, issue: 2 },
+      { id: 3, issue: 1 },
+      { id: 4, issue: 2 },
+      { id: 5, issue: 3 },
+      { id: 6, issue: 3 }
+    ]
+
+    async function emitSubscribe (...args) {
+      socket.emit(...args)
+      const { events } = await pEvent(socket, 'events', {
```
Comment (petersutter) [inline backend/test/acceptance/io.spec.js]: iterator

File: backend/server.js (hunk 1 above)
Comment (holgerkoser) [reply backend/server.js]: ✔️

File: backend/lib/server.js (hunk 2 above)
Comment (holgerkoser) [reply backend/lib/server.js]: It is needed for the test. I have moved the code into the test.
✔️

File: backend/lib/watches/journals.js (hunk 3 above)
Comment (holgerkoser) [reply backend/lib/watches/journals.js]: Added comment and set to undefined

File: backend/test/support/nocks/k8s.js (hunk 4 above)
Comment (holgerkoser) [reply backend/test/support/nocks/k8s.js]: ✔️

File: backend/test/services.terminals.spec.js (hunk 5 above)
Comment (holgerkoser) [reply backend/test/services.terminals.spec.js]: ✔️

File: backend/test/services.terminals.spec.js (hunk 6 above)
Comment (holgerkoser) [reply backend/test/services.terminals.spec.js]: ✔️

File: backend/test/services.terminals.spec.js (hunk 7 above)
Comment (holgerkoser) [reply backend/test/services.terminals.spec.js]: ✔️

File: backend/test/utils.spec.js (hunk 8 above)
Comment (holgerkoser) [reply backend/test/utils.spec.js]: ✔️

File: backend/test/watches.spec.js (hunk 9 above)
Comment (holgerkoser) [reply backend/test/watches.spec.js]: ✔️

File: backend/test/watches.spec.js (hunk 12 above)
Comment (holgerkoser) [reply backend/test/watches.spec.js]: ✔️

File: backend/test/watches.spec.js (hunk 11 above)
Comment (holgerkoser) [reply backend/test/watches.spec.js]: ✔️

File: backend/test/watches.spec.js (hunk 13 above)
Comment (holgerkoser) [reply backend/test/watches.spec.js]: ✔️

File: backend/test/watches.spec.js (hunk 14 above)
Comment (holgerkoser) [reply backend/test/watches.spec.js]: ✔️

File: backend/test/acceptance/io.spec.js (hunk 15 above)
Comment (holgerkoser) [reply backend/test/acceptance/io.spec.js]: ✔️

File: backend/lib/kubernetes-client/debug.js [hunk 16]
```diff
@@ -37,9 +37,24 @@ function afterResponse (response) {
...
+    logger.error('URI parse error:', err)
+  }
+}
+
 function beforeRequest (options) {
-  const { href, method, headers, body, key, cert } = options
-  const uri = new URL(href)
+  const { url, href, method, headers, body, key, cert } = options
+  let uri
+  if (url instanceof URL) {
```
Comment (petersutter) [inline backend/lib/kubernetes-client/debug.js]: assertion test

File: backend/test/healthz.spec.js [hunk 17]
```diff
@@ -35,14 +35,14 @@ describe('healthz', function () {
   })
 
   it('should successfully check health', async function () {
-    await healthCheck()
+    await healthCheck(true)
```
Comment (petersutter) [inline backend/test/healthz.spec.js]: is there a test for healthz(false)?

File: backend/test/services.terminals.spec.js [hunk 18]
```diff
@@ -60,6 +60,9 @@ describe('services', function () {
       core: {
         secrets: {
           list (namespace, query) {
+            expect(query).to.eql({
+              labelSelector: 'runtime=gardenTerminalHost'
```
Comment (petersutter) [inline backend/test/services.terminals.spec.js]: use spy and check for namespace=garden

File: backend/lib/kubernetes-client/debug.js (hunk 16 above)
Comment (holgerkoser) [reply backend/lib/kubernetes-client/debug.js]: ✔️

File: backend/test/healthz.spec.js (hunk 17 above)
Comment (holgerkoser) [reply backend/test/healthz.spec.js]: ✔️

File: backend/test/services.terminals.spec.js (hunk 18 above)
Comment (holgerkoser) [reply backend/test/services.terminals.spec.js]: ✔️

File: backend/lib/logger/Logger.js [hunk 19]
```diff
@@ -49,11 +49,11 @@ class Logger {
     }
   }
 
-  request ({ id, uri, method, httpVersion = '1.1', user, headers, body }) {
+  request ({ id, url, method, httpVersion = '1.1', user, headers, body }) {
     if (!this.isDisabled(LEVELS.trace + 1)) {
```
Comment (petersutter) [inline backend/lib/logger/Logger.js]: debug level and for response and log timestamp

File: backend/test/services.terminals.spec.js [hunk 20]
```diff
@@ -541,9 +546,17 @@ describe('services', function () {
           disabled: false,
           gardenTerminalHostDisabled: false
         }
+        const listSecretsSpy = sandbox.spy(client.core.secrets, 'list')
         createConfigStub({ gardenTerminalHost, bootstrap })
         const bootstrapper = new Bootstrapper()
         await pEvent(bootstrapper, 'empty')
+        expect(listSecretsSpy).to.be.calledOnce
```
Comment (petersutter) [inline backend/test/services.terminals.spec.js]: rm

Comment (holgerkoser) [reply backend/test/services.terminals.spec.js]: ✔️

File: backend/lib/logger/Logger.js (hunk 19 above)
Comment (holgerkoser) [reply backend/lib/logger/Logger.js]: ✔️

**Pull request merged**